components/eamxx/src/diagnostics/tests/field_at_height_tests.cpp

```diff
@@ -120,106 +143,181 @@ TEST_CASE("field_at_height")
-        auto fi = v_int.subfield(0,i);
-        auto di = d.subfield(0,i);
-        auto tgt = fi.subfield(1,zp1).clone();
-        tgt.update(fi.subfield(1,zm1),0.5,0.5);
-        REQUIRE (views_are_equal(di,tgt,&comm));
+      {
+        print("    -> scalar interface field...............\n");
+        auto d = run_diag (s_int,ref_int,loc,surf_ref);
+        auto tgt = get_ref_field(zm1+1,zp1+1,s_int,surf_ref);
+        REQUIRE (views_are_equal(d,tgt,&comm));
+        print("    -> scalar interface field............... OK!\n");
+      }
+      {
+        print("    -> vector midpoint field...............\n");
+        auto d = run_diag (v_mid,ref_mid,loc,surf_ref);
+        // We can't subview over 3rd index and keep layout right,
+        // so do all cols separately
+        for (int i=0; i<ncols; ++i) {
+          auto fi = v_mid.subfield(0,i);
+          auto di = d.subfield(0,i);
+	  // Because we are already cycling through cols we don't need
+	  // get_ref_field to adjust for above sealevel, we do it locally.
+	  auto zm1_adj = surf_ref == "surface" ? zm1 :
+		  ( i + zm1 >= nlevs ? nlevs-1 : i + zm1 );
+	  auto zp1_adj = surf_ref == "surface" ? zp1 :
+		  ( i + zp1 >= nlevs ? nlevs-1 : i + zp1 );
+          auto tgt = get_ref_field(zm1_adj,zp1_adj,fi,"");
+          REQUIRE (views_are_equal(di,tgt,&comm));
+        }
+        print("    -> vector midpoint field............... OK!\n");
       }
-      print("    -> vector interface field............... OK!\n");
+      {
+        print("    -> vector interface field...............\n");
+        auto d = run_diag (v_int,ref_int,loc,surf_ref);
+        // We can't subview over 3rd index and keep layout right,
+        // so do all cols separately
+        for (int i=0; i<ncols; ++i) {
+          auto fi = v_int.subfield(0,i);
+          auto di = d.subfield(0,i);
+	  // Because we are already cycling through cols we don't need
+	  // get_ref_field to adjust for above sealevel, we do it locally.
+	  auto zm1_adj = surf_ref == "surface" ? zm1 :
+		  ( i + zm1 >= nlevs ? nlevs-1 : i + zm1 );
+	  auto zp1_adj = surf_ref == "surface" ? zp1 :
+		  ( i + zp1 >= nlevs ? nlevs-1 : i + zp1 );
+          auto tgt = get_ref_field(zm1_adj+1,zp1_adj+1,fi,"");
+          REQUIRE (views_are_equal(di,tgt,&comm));
+        }
+        print("    -> vector interface field............... OK!\n");
+      }
+    }
+  }
+}
+
+//-----------------
+Field get_ref_field(const int lev, const Field& src_data, const std::string& surf_ref)
+{
+  // A single level implies this is a test where the height matches a single level.
+  if (surf_ref == "sealevel") {
+    auto tgt = src_data.subfield(1,lev).clone();
+    const auto& tgt_view = tgt.get_view<Real*,Host>();
+    const auto dims = src_data.get_header().get_identifier().get_layout().dims();
+    const auto src_view = src_data.get_view<const Real**,Host>();
+    for (int i = 0; i < dims[0]; ++i) {
+      int ilev = lev + i >= dims[1] ? dims[1]-1 : lev + i;
```

Comment:
I'm confused by this line. The index `i` is over columns, while here you use it to choose at what level to subview the input field?